apps/backend/src/users/users.controller.ts

```diff
@@ -71,8 +75,27 @@ export class UsersController {
   @Post()
   @UsePipes(new PasswordsMatchPipe(), new PasswordComplexityPipe())
   @UseFilters(new UniqueConstraintErrorFilter())
-  async create(@Body() createUserDto: CreateUserDto): Promise<UserDto> {
-    return new UserDto(await this.usersService.create(createUserDto));
+  @UseGuards(ImplicitAllowJwtAuthGuard)
+  async create(
+    @Body() createUserDto: CreateUserDto,
+    @Request() request: {user?: User}
+  ): Promise<UserDto> {
+    if (this.configService.isRegistrationAllowed()) {
+      return new UserDto(await this.usersService.create(createUserDto));
+    } else {
+      if (request.user) {
+        const abac = this.authz.abac.createForUser(request.user);
+        ForbiddenError.from(abac)
+          .setMessage(
+            'User registration is disabled. Please ask your system administrator to create the account.'
+          )
+          .throwUnlessCan(Action.ReadAll, User);
+        return new UserDto(await this.usersService.create(createUserDto));
+      }
+      throw new ForbiddenException(
+        'User registration is disabled. Please ask your system administrator to create the account.'
+      );
```

Comment:
Can this be restructured to not duplicate the call to `usersService.create` and the `ForbiddenException`?